lib/ansible/modules/crypto/openssl_certificate.py

```diff
@@ -0,0 +1,764 @@
+            module.params['state'],
+            module.params['force'],
+            module.check_mode
+        )
+
+        self.provider = module.params['provider']
+        self.privatekey_path = module.params['privatekey_path']
+        self.privatekey_passphrase = module.params['privatekey_passphrase']
+        self.csr_path = module.params['csr_path']
+        self.cert = None
+        self.privatekey = None
+        self.module = module
+
+    def check(self, module, perms_required=True):
+        """Ensure the resource is in its desired state."""
+
+        state_and_perms = super(Certificate, self).check(module, perms_required)
+
+        def _validate_privatekey():
+            if self.privatekey_path:
+                ctx = OpenSSL.SSL.Context(OpenSSL.SSL.TLSv1_2_METHOD)
+                ctx.use_privatekey(self.privatekey)
+                ctx.use_certificate(self.cert)
+                try:
+                    ctx.check_privatekey()
+                    return True
+                except OpenSSL.SSL.Error:
+                    return False
+
+        if not state_and_perms:
+            return False
+
+        self.cert = crypto_utils.load_certificate(self.path)
+
+        if self.privatekey_path:
+            self.privatekey = crypto_utils.load_privatekey(
+                self.privatekey_path,
+                self.privatekey_passphrase
+            )
+            return _validate_privatekey()
+
+        return True
+
+
+class SelfSignedCertificate(Certificate):
+    """Generate the self-signed certificate."""
+
+    def __init__(self, module):
+        super(SelfSignedCertificate, self).__init__(module)
+        self.serial_number = randint(1000, 99999)
+        self.notBefore = module.params['selfsigned_notBefore']
+        self.notAfter = module.params['selfsigned_notAfter']
+        self.digest = module.params['selfsigned_digest']
+        self.csr = crypto_utils.load_certificate_request(self.csr_path)
+        self.privatekey = crypto_utils.load_privatekey(
+            self.privatekey_path, self.privatekey_passphrase
+        )
+        self.cert = None
+
+    def generate(self, module):
+
+        if not os.path.exists(self.privatekey_path):
+            raise CertificateError(
+                'The private key %s does not exist' % self.privatekey_path
+            )
+
+        if not os.path.exists(self.csr_path):
+            raise CertificateError(
+                'The certificate signing request file %s does not exist' % self.csr_path
+            )
+
+        if not self.check(module, perms_required=False) or self.force:
+            cert = crypto.X509()
+            cert.set_serial_number(self.serial_number)
+            if self.notBefore:
+                cert.set_notBefore(self.notBefore)
+            else:
+                cert.gmtime_adj_notBefore(0)
+            if self.notAfter:
+                cert.set_notAfter(self.notAfter)
+            else:
+                cert.gmtime_adj_notAfter(315360000)
```

Comment:
Purely cosmetic, but I tend to prefer constant like this to be expanded, or comment to say "that's 10 years in second"